File: ShopGuard/ShopGuard.E2ETests/Pages/LoginPage.cs
using Microsoft.Playwright;

namespace ShopGuard.E2ETests.Pages;

/// <summary>
/// Page object for /auth/login of the Toolshop demo.
/// </summary>
public sealed class LoginPage(IPage page) : BasePage(page)
{
    private ILocator EmailInput => Page.Locator("[data-test='email']");
    private ILocator PasswordInput => Page.Locator("[data-test='password']");
    private ILocator LoginButton => Page.Locator("[data-test='login-submit']");
    private ILocator LoginError => Page.Locator("[data-test='login-error']");
    private ILocator EmailFieldError => Page.Locator("[data-test='email-error']");

    public Task OpenAsync() => NavigateAsync("auth/login");

    public async Task LoginAsync(string email, string password)
    {
        await EmailInput.FillAsync(email);
        await PasswordInput.FillAsync(password);
        await LoginButton.ClickAsync();
    }

    /// <summary>Returns the summary error shown for rejected credentials.</summary>
    public async Task<string> GetErrorMessageAsync()
    {
        await LoginError.WaitForAsync();
        return await LoginError.InnerTextAsync();
    }

    /// <summary>Returns the inline validation error below the e-mail field, if any.</summary>
    public async Task<string> GetEmailValidationErrorAsync()
        => await EmailFieldError.IsVisibleAsync()
            ? await EmailFieldError.InnerTextAsync()
            : string.Empty;
}

File: playwright-ts/pages/login-page.ts
import { Locator, Page } from '@playwright/test';

/**
 * Page object for /auth/login of the Toolshop demo.
 * Mirrors the C# LoginPage so both worlds stay structurally identical.
 */
export class LoginPage {
  readonly loginError: Locator;
  readonly userNavMenu: Locator;

  constructor(private readonly page: Page) {
    this.loginError = page.locator("[data-test='login-error']");
    this.userNavMenu = page.locator("[data-test='nav-menu']");
  }

  async open(): Promise<void> {
    await this.page.goto('/auth/login');
  }

  async login(email: string, password: string): Promise<void> {
    await this.page.locator("[data-test='email']").fill(email);
    await this.page.locator("[data-test='password']").fill(password);
    await this.page.locator("[data-test='login-submit']").click();
  }
}
